### PR TITLE
[DUOS-512][risk=low] User Flow Testing - Researcher DAR creation

### PR DESCRIPTION
**Pull request by rfricke-asymmetrik**
**Addresses**
https://broadinstitute.atlassian.net/browse/DUOS-512 

**Changes**
Adds more scenarios to the gatling implementation. The specific scenarios added in this PR are for a researcher completing the process of logging in, selecting two data sets and completing the data access request application process.

This adds spray.json to the scala project for json parsing and adds a chains structure that allows for reuse of different ChainBuilder objects across different scenarios. It also changes the existing http request functions in Requests.scala to HttpRequestBuilder objects instead of ChainBuilder.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/DataBiosphere/consent/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing
- [ ] I've updated Swagger to reflect any API changes

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/configs/application.conf.ctmpl [hunk 1]
```diff
@@ -1,7 +1,9 @@
 {{with $environment := env "ENVIRONMENT"}}
 
 consent {
-  baseUrl = {{if eq $environment "local"}}"https://local.broadinstitute.org:27443/"{{else}}"https://consent.dsde-{{$environment}}.broadinstitute.org/"{{end}}
+  baseUrl = {{if eq $environment "local"}}"https://local.broadinstitute.org:27443/"{{else}}"https://consent.dsde-{{$environment}}.broadinstitute.org/"{{end}},
+  fireCloudUrl = "https://firecloud-orchestration.dsde-{{$environment}}.broadinstitute.org/",
+  profileUrl = "https://profile-dot-broad-shibboleth-prod.appspot.com/{{$environment}}"
```
Comment (rushtong) [inline automation/configs/application.conf.ctmpl]: There are really only two values here, dev or prod. All non-prod envs should use `dev`, and we should never be running this against prod, so let's hard code this to dev.
```suggestion
  profileUrl = "https://profile-dot-broad-shibboleth-prod.appspot.com/dev"
```

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            Requests.Admin.initConsole(200, additionalHeaders)
+        )
+    }
+
+    def manageAccess(additionalHeaders: Map[String, String]): ChainBuilder = {
+        exec(
+            Requests.Dar.manageDar(200, additionalHeaders)
+        )
+        .exec { session =>
+            implicit val darManageFormat = JsonProtocols.DataAccessRequestManageFormat
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: nit: prefer explicit typing, here and throughout:
```suggestion
            implicit val darManageFormat: JsonProtocols.DataAccessRequestManageFormat.type = JsonProtocols.DataAccessRequestManageFormat
```

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            implicit val userFormat = JsonProtocols.userFormat
+            implicit val dacFormat = JsonProtocols.dacFormat
+            implicit val electionStatusFormat = JsonProtocols.electionStatusFormat
+
+            val manageDarStr = session(Requests.Dar.manageDarResponse).as[String]
+            var manageDars = manageDarStr.parseJson.convertTo[Seq[DataAccessRequestManage]]
+
+            val newManageDars = DarService.setManageRolesByOwner(manageDars)
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: this looks unused

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            Requests.Dar.manageDar(200, additionalHeaders)
+        )
+        .exec { session =>
+            implicit val darManageFormat = JsonProtocols.DataAccessRequestManageFormat
+            implicit val userFormat = JsonProtocols.userFormat
+            implicit val dacFormat = JsonProtocols.dacFormat
+            implicit val electionStatusFormat = JsonProtocols.electionStatusFormat
+
+            val manageDarStr = session(Requests.Dar.manageDarResponse).as[String]
+            var manageDars = manageDarStr.parseJson.convertTo[Seq[DataAccessRequestManage]]
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: nit: val

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+            val newManageDars = DarService.setManageRolesByOwner(manageDars)
+
+            val researcherDars = DarService.getPendingDARsByMostRecent(manageDars, 2)
+            val electionStatus = ElectionStatus("Open", false)
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: nit: name parameters:
```suggestion
            val electionStatus = ElectionStatus("Open", finalAccessVote = false)
```

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DarChains.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    def finalDarSubmit(additionalHeaders: Map[String, String]): ChainBuilder = {
+        exec { session =>
+            implicit val researchFormat = JsonProtocols.researcherPropertyFormat
+            implicit val researcherFormat = JsonProtocols.ResearcherInfoFormat
+            implicit val dataAccessRequestDataFormat = JsonProtocols.DataAccessRequestDataFormat
+            implicit val dataAccessFormat = JsonProtocols.dataAccessRequestFormat
+            implicit val dataSetFormat = JsonProtocols.dataSetFormat
+            implicit val dataSetEntryFormat = JsonProtocols.dataSetEntryFormat
+            var dataMap = collection.mutable.Map[String, JsValue]()
+            var darMap = collection.mutable.Map[String, JsValue]()
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DarChains.scala]: these look unused

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def dataSetCatalogPickTwo(additionalHeaders: Map[String, String]): ChainBuilder = {
+        exec(
+            Requests.User.dataSetCatalog(200, "${dacUserId}", additionalHeaders)
+        )
+        .exec { session =>
+            implicit val dataSetFormatJson = JsonProtocols.dataSetFormat
+            implicit val dataSetEntryJson = JsonProtocols.dataSetEntryFormat
+            implicit val dataAccessRequestJson = JsonProtocols.dataAccessRequestDraftFormat
+
+            var dataSetStr = session(Requests.DataSet.dataSetResponse).as[String]
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: nit: val

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            val dataSets = dataSetStr.parseJson.convertTo[Seq[DataSet]]
+            
+            var chosenSets = dataSets
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: nit: val

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                .groupBy(_.dacId)
+                .map(_._2.head)
+                .take(2)
+
+            var setIds = new ListBuffer[Int]()
+            var entrySets = new ListBuffer[DataSetEntry]()
+            chosenSets.foreach(set => {
+                setIds += set.dataSetId
+                entrySets += DataSetEntryBuilder.fromDataSet(set)
+            })
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: When you find yourself mutating vars in scala, there is almost always a more functional non-mutating way to do it. For example:
```
            val setIds: Seq[Int] = chosenSets.map(s => s.dataSetId).toSeq
            val entrySets: Seq[DataSetEntry] = chosenSets.map(s => DataSetEntryBuilder.fromDataSet(s)).toSeq
```
We strongly prefer immutable vals over their mutable counterparts.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+import org.broadinstitute.dsp.consent.models.DataAccessRequestModels._
+import org.broadinstitute.dsp.consent.models.NihModels._
+import org.broadinstitute.dsp.consent.models.DacModels._
+import org.broadinstitute.dsp.consent.models.ElectionModels._
+
+
+import scala.util.{Failure, Success, Try}
+
+object JsonProtocols extends DefaultJsonProtocol {
+    implicit val researcherPropertyFormat = jsonFormat4(ResearcherProperty)
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: nit: more type annotation required here

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            case None => default
+        }
+    }
+
+    implicit object DataAccessRequestManageFormat extends JsonFormat[DataAccessRequestManage] {
+        def write(darm: DataAccessRequestManage) = {
+            var map = collection.mutable.Map[String, JsValue]()
+            val manualList = List("errors")
+            darm.getClass.getDeclaredFields
+                .filterNot(f => manualList.exists(_ == f.getName))
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: nit: This is a bit simpler:
```suggestion
                .filterNot(f => manualList.contains(f.getName))
```

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+                        case Some(y: String) => map += f.getName -> y.toJson
+                        case Some(l: Long) => map += f.getName -> l.toJson
+                        case Some(i: Int) => map += f.getName -> i.toJson
+                        case Some(u: User) => map += f.getName -> u.toJson
+                        case Some(d: Dac) => map += f.getName -> d.toJson
+                        case _ => map += f.getName -> JsNull
+                    }
+                }
+
+            if (!darm.errors.isEmpty)
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: nit: `if (darm.errors.isDefined)` is slightly safer here

File: automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala [hunk 13]
```diff
@@ -4,49 +4,52 @@ import java.net.URLEncoder
 
 import io.gatling.core.Predef._
 import io.gatling.core.structure.ChainBuilder
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala]: This looks unused now

File: automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala [hunk 14]
```diff
@@ -4,49 +4,52 @@ import java.net.URLEncoder
...
+import io.gatling.http.request.builder._
 import io.gatling.http.Predef._
 import org.broadinstitute.dsp.consent.TestConfig
+import org.broadinstitute.dsp.consent.models._
+import spray.json._
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala]: Haven't tested, but these look unused

File: automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/AdminScenarios.scala [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import io.gatling.http.Predef._
+import org.broadinstitute.dsp.consent.{TestConfig, TestRunner}
+import org.broadinstitute.dsp.consent.requests.Requests
+import scala.concurrent.duration._
+import spray.json._
+import DefaultJsonProtocol._
+import org.broadinstitute.dsp.consent.services._
+import scala.collection.mutable.ListBuffer
+import scala.collection.immutable.IntMap
+import org.broadinstitute.dsp.consent.chains.{DarChains, DataSetChains, NihChains, AdminChains}
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/AdminScenarios.scala]: quite a bit of unused imports here

File: automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/UserScenarios.scala [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import org.broadinstitute.dsp.consent.{TestConfig, TestRunner}
+import org.broadinstitute.dsp.consent.requests.Requests
+import scala.concurrent.duration._
+import spray.json._
+import DefaultJsonProtocol._
+import org.broadinstitute.dsp.consent.models._
+import org.broadinstitute.dsp.consent.services._
+import scala.collection.mutable.ListBuffer
+import scala.collection.immutable.IntMap
+import org.broadinstitute.dsp.consent.chains.{DarChains, DataSetChains, NihChains, AdminChains}
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/UserScenarios.scala]: quite a bit of unused imports here

File: automation/src/test/scala/org/broadinstitute/dsp/consent/services/DarService.scala [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        dataMap += ("oneGender" -> JsBoolean(false))
+        dataMap += ("pediatric" -> JsBoolean(false))
+        dataMap += ("illegalBehavior" -> JsBoolean(false))
+        dataMap += ("addiction" -> JsBoolean(false))
+        dataMap += ("sexualDiseases" -> JsBoolean(false))
+        dataMap += ("stigmatizedDiseases" -> JsBoolean(false))
+        dataMap += ("vulnerablePopulation" -> JsBoolean(false))
+        dataMap += ("populationMigration" -> JsBoolean(false))
+        dataMap += ("psychiatricTraits" -> JsBoolean(false))
+        dataMap += ("notHealth" -> JsBoolean(false))
```
Comment (rushtong) [inline automation/src/test/scala/org/broadinstitute/dsp/consent/services/DarService.scala:84]: I think this is an example of a mutability that makes sense 👍🏽

**Review by rushtong — CHANGES_REQUESTED** [review]
Lots of great work here, thank you so much for this! Please see feedback inline, but in general, my feedback is more along the lines of stylistic scala
* Type annotations for vals is important
* Use vals instead of vars
* Avoid mutation as much as it makes sense to

File: automation/src/test/scala/org/broadinstitute/dsp/consent/services/DarService.scala (hunk 17 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/services/DarService.scala:84]: Thank you 👍

File: automation/configs/application.conf.ctmpl (hunk 1 above)
Comment (rfricke-asymmetrik) [reply automation/configs/application.conf.ctmpl]: Makes sense. I went ahead and hardcoded all three urls to use dev.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala (hunk 2 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: I updated the typings everywhere I could see where it was missing.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/UserScenarios.scala (hunk 16 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/UserScenarios.scala]: Leftover imports from before a code restructuring. I have removed the unecessary imports

File: automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/AdminScenarios.scala (hunk 15 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/scenarios/AdminScenarios.scala]: Leftovers removed.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala (hunk 14 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala]: Leftovers removed

File: automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala (hunk 13 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/requests/Requests.scala]: removed

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala (hunk 5 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: There were a few spots where name parameters were not used. I updated those as well as this one for consistency.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DarChains.scala (hunk 6 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DarChains.scala]: leftovers from restructuring, removed

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala (hunk 9 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: Thank you. Took this example and also updated another similar code block elsewhere.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala (hunk 11 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: Took this suggestion as well and applied it everywhere that exists was used in this fashion.

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala (hunk 12 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: Changed isEmpty to isDefined and removed the bang in all places like this

File: automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala (hunk 10 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/models/JsonProtocols.scala]: Types added everywhere

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala (hunk 8 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: noted, var changed to val for both of these variables

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala (hunk 7 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/DataSetChains.scala]: updated

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala (hunk 4 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: updated

File: automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala (hunk 3 above)
Comment (rfricke-asymmetrik) [reply automation/src/test/scala/org/broadinstitute/dsp/consent/chains/AdminChains.scala]: Variable was created and supposed to be passed into the next service function. This is resolved now.

**Review by rushtong — APPROVED** [review]
Thank you for the changes, great work here!

**Pull request merged**